plugins/modules/ec2_transit_gateway.py

```diff
@@ -269,71 +255,63 @@ def process(self):
-        transit_gateway = dict()
-
-        while wait_timeout > time():
-            try:
-                transit_gateway = self.get_matching_tgw(tgw_id=tgw_id, skip_deleted=skip_deleted)
-
-                if transit_gateway:
-                    if self._check_mode:
-                        transit_gateway["state"] = status
+        max_attempts = wait_timeout // polling_increment_secs
 
-                    if transit_gateway.get("state") == status:
-                        status_achieved = True
-                        break
+        waiter_method = f"transit_gateway_{status}"
 
-                    elif transit_gateway.get("state") == "failed":
-                        break
-
-                else:
-                    sleep(polling_increment_secs)
-
-            except ClientError as e:
-                self._module.fail_json_aws(e)
+        try:
+            wait_for_resource_state(
+                self._connection,
+                self._module,
+                waiter_method,
+                TransitGatewayIds=[tgw_id],
+                delay=polling_increment_secs,
+                max_attempts=max_attempts,
+            )
+        except AnsibleEC2Error as e:
+            self._module.fail_json_aws_error(e)
 
-        if not status_achieved:
-            self._module.fail_json(msg="Wait time out reached, while waiting for results")
+        transit_gateway = self.get_matching_tgw(tgw_id=tgw_id, skip_deleted=skip_deleted)
+        if transit_gateway is None:
+            self._module.fail_json(msg="Transit Gateway not found after waiting.")
 
         return transit_gateway
 
-    def get_matching_tgw(self, tgw_id, description=None, skip_deleted=True):
-        """search for  an existing tgw by either tgw_id or description
-        :param tgw_id:  The AWS id of the transit gateway
-        :param description:  The description of the transit gateway.
-        :param skip_deleted: ignore deleted transit gateways
-        :return dict: transit gateway object
+    def get_matching_tgw(
+        self, tgw_id: Optional[str], description: Optional[str] = None, skip_deleted: bool = True
+    ) -> Optional[Dict[str, Any]]:
+        """Search for an existing tgw by either tgw_id or description.
+        :param tgw_id: The AWS id of the transit gateway.
+        :param description: The description of the transit gateway.
+        :param skip_deleted: Ignore deleted transit gateways.
+        :return: Transit gateway object.
         """
         filters = []
+        params = {}
         if tgw_id:
             filters = ansible_dict_to_boto3_filter_list({"transit-gateway-id": tgw_id})
 
-        try:
-            response = AWSRetry.exponential_backoff()(self._connection.describe_transit_gateways)(Filters=filters)
-        except (ClientError, BotoCoreError) as e:
-            self._module.fail_json_aws(e)
+        params["Filters"] = filters
+        response = describe_ec2_transit_gateways(self._connection, **params)
```

Comment:
```suggestion
        response = describe_ec2_transit_gateways(self._connection, Filters=filters)
```

Reply:
I think we should keep it to `**params` as `describe_ec2_transit_gateways` from a.aws.module_utils/ec2 is used by both `ec2_transit_gateway` and `ec2_transit_gateway_info` with two separate calls. One only sends Filters, while other sends Filters and TransitGatewayIds

```
        filters = []
        params = {}
        if tgw_id:
            filters = ansible_dict_to_boto3_filter_list({"transit-gateway-id": tgw_id})

        params["Filters"] = filters
        response = describe_ec2_transit_gateways(self._connection, **params)
```
and
```
    params = {}
    if transit_gateway_ids:
        params["TransitGatewayIds"] = transit_gateway_ids
    if filters:
        params["Filters"] = filters

    result = describe_ec2_transit_gateways(connection, **params)
```

Reply:
resolving for now, please reopen if required.